File: nii2png.py
#!/usr/bin/python3.7
'''''''''''''''''''''''''''''''''''''''''
'       nii2png for Python 3.7          '
'         NIfTI Image Converter         '
'                v0.2.7                 '
'                                       '
'     Written by Alexander Laurence     '
' http://Celestial.Tokyo/~AlexLaurence/ '
'    alexander.adamlaurence@gmail.com   '
'              09 May 2019              '
'              MIT License              '
'''''''''''''''''''''''''''''''''''''''''

import scipy, numpy, shutil, os, nibabel
import sys, getopt

def main(argv):
        
        inputfile = ''
        outputfile = ''
        try:
                opts, args = getopt.getopt(argv,"hi:o:",["ifile=","ofile="])
        except getopt.GetoptError:
                print('nii2png.py -i <inputfile> -o <outputfile>')
                sys.exit(2)
        for opt, arg in opts:
                if opt == '-h':
                        print('nii2png.py -i <inputfile> -o <outputfile>')
                        sys.exit()
                elif opt in ("-i", "--input"):
                        inputfile = arg
                elif opt in ("-o", "--output"):
                        outputfile = arg
        print('Input file is ', inputfile)
        print('Output folder is ', outputfile)

        # set fn as your 4d nifti file
        image_array = nibabel.load(inputfile).get_data()
        print(len(image_array.shape))
        
        # set rotations
        data_4d = image_array[:, :, i, j]
        data_rot1_4d = numpy.rot90(image_array[:, :, i, j])
        data_rot2_4d = numpy.rot90(data_rot1_4d)
        data_rot3_4d = numpy.rot90(data_rot2_4d)
        
        data_3d = image_array[:, :, i]
        data_rot1_3d = numpy.rot90(image_array[:, :, i])
        data_rot2_3d = numpy.rot90(data_rot1_3d)
        data_rot3_3d = numpy.rot90(data_rot2_3d)
        
        ask_rotate = input('Would you like to rotate the orientation? (y/n) ')
        
        if ask_rotate.lower() == 'y':
                ask_rotate_num = input('OK. By 90° 180° or 270°? ')
                
                if ask_rotate_num == 90:
                        if len(image_array.shape) == 4:
                                data = data_rot1_4d
                        elif len(image_array.shape) == 3:
                                data = data_rot1_3d
                elif ask_rotate_num == 180:
                        if len(image_array.shape) == 4:
                                data = data_rot2_4d
                        elif len(image_array.shape) == 3:
                                data = data_rot2_3d
                elif ask_rotate_num == 280: 
                        if len(image_array.shape) == 4:
                                data = data_rot3_4d
                        elif len(image_array.shape) == 3:
                                data = data_rot3_3d
                else:
                        print('Sorry, I did not understand that. Quitting...')
                        sys.exit()
        elif ask_rotate.lower() == 'n':
                print('OK, I will convert it as it is.')
                if len(image_array.shape) == 4:
                        data = data_4d
                elif len(image_array.shape) == 3:
                        data = data_3d
        else:
                print('Sorry, I did not understand that. Quitting...')
                sys.exit()

        # if 4D image inputted
        if len(image_array.shape) == 4:
                # set 4d array dimension values
                nx, ny, nz, nw = image_array.shape

                # set destination folder
                if not os.path.exists(outputfile):
                        os.makedirs(outputfile)
                        print("Created ouput directory: " + outputfile)

                print("Converting NIfTI to png...")

                volumes = image_array.shape[3]
                slices = image_array.shape[2]

                volume_counter = 0
                image_counter = 0

                # iterate through volumes
                for j in range(0, volumes):
                  slice_counter = 0
                  # iterate through slices
                  for i in range(0, slices):
                    #alternate slices and save as png
                    if (slice_counter % 1) == 0:
                      image_name = inputfile[:-4] + "_t" + "{:0>3}".format(str(j)) + "_z" + "{:0>3}".format(str(i))+ ".png"
                      scipy.misc.imsave(image_name, data)

                      #move images to folder
                      src = image_name
                      shutil.move(src, outputfile)
                      slice_counter += 1
                      image_counter += 1
                  volume_counter += 1

                print("Finished converting images") 

        # else if 3D image inputted
        elif len(image_array.shape) == 3:
                # set 4d array dimension values
                nx, ny, nz = image_array.shape

                # set destination folder
                if not os.path.exists(outputfile):
                    os.makedirs(outputfile)
                    print("Created ouput directory: " + outputfile)

                print("Converting NIfTI to png...")

                slices = image_array.shape[2]

                image_counter = 0
                slice_counter = 0
                # iterate through slices
                for i in range(0, slices):
                        # set nifti as a numpy array
                        data = numpy.rot90(image_array[:, :, i])
                        #alternate slices and save as png
                        if (slice_counter % 1) == 0:
                                image_name = inputfile[:-4] + "_z" + "{:0>3}".format(str(i))+ ".png"
                                scipy.misc.imsave(image_name, data)

                                #move images to folder
                                src = image_name
                                shutil.move(src, outputfile)
                                slice_counter += 1
                                image_counter += 1

                print("Finished converting images") 
        else:
                print("Not a 3D or 4D Image. Please try again.")

# call the function to start the program
if __name__ == "__main__":
   main(sys.argv[1:])

File: nii2png.m
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%       nii2png for Matlab R2017b       %
%         NIfTI Image Converter         %
%                v0.0.1                 %
%                                       %
%     Written by Alexander Laurence     %
% http://Celestial.Tokyo/~AlexLaurence/ %
%    alexander.adamlaurence@gmail.com   %
%              09 May 2019              %
%              MIT License              %
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

% Set the Working Directory
clear; warning off; current = pwd;
path = uigetdir(pwd,'select your working directory');
cd(path);

% Select NIfTI
[file,path] = uigetfile('*.nii');
if isequal(file,0)
   disp('User selected Cancel');
else
   disp(['User selected ', fullfile(path,file)]);
   
   % Read NIfTI Data and Header Info
   image = niftiread(fullfile(path,file));
   image_info = niftiinfo(fullfile(path,file));
   nifti_array = size(image);

   % If this is a 4D NIfTI
   if length(nifti_array) == 4
       
       % Create output folder
       mkdir png
       
       % Get Vols and Slice
       volumes = nifti_array(4);
       slices = nifti_array(3);      
       
       disp('Converting NIfTI to png, please wait...')
       % Iterate Through Vol
       j = 1 ;
       while j <= volumes
           slice_counter = 0;
            % Iterate Through Slices
            i = 1;
            while i <= slices
                % Alternate Slices
                if mod(slice_counter, 1) == 0        
                    % Set Filename as per slice and vol info
                    filename = file(1:end-4) + "_t" + sprintf('%03d', j) + "_z" + sprintf('%03d', i) + ".png";
                    
                    % Convert Image to Double
                    data = im2double(image);
                    
                    % Write Image
                    imwrite(rot90(mat2gray(data(:,:,i,j))), char(filename));
                                        
                    % If we reached the end of the slices
                    if i == slices
                        % But not the end of the volumes
                        if j < volumes
                            % Move to the next volume                            
                            j = j + 1;
                            % Write the image
                            imwrite(rot90(mat2gray(data(:,:,i,j))), char(filename));
                        % Else if we reached the end of slice and volume
                        else         
                            % Write Image
                            imwrite(rot90(mat2gray(data(:,:,i,j))), char(filename));
                            disp('Finished!')
                            return
                        end
                    end
                 
                    % Move Images To Folder
                    movefile(char(filename),'png');
                    
                    % Increment Counters
                    slice_counter = slice_counter + 1;
                    
                    percentage = strcat('Please wait. Converting...', ' ', num2str((j/volumes)*100), '% Complete');
                    
                    if ((j/volumes)*100) == 100
                        disp('100% Complete!');
                    else
                        disp(percentage);
                    end                    
                end
                i = i + 1;
            end
       j = j + 1;
       end
   % Else if this is a 3D NIfTI
   elseif length(nifti_array) == 3
       % Create output folder
       mkdir png
       
       % Get Vols and Slice
       slices = nifti_array(3);      
       
       disp('Converting NIfTI to png, please wait...')

       slice_counter = 0;
        % Iterate Through Slices
        i = 1;
        while i <= slices
            % Alternate Slices
            if mod(slice_counter, 1) == 0        
                % Set Filename as per slice and vol info
                filename = file(1:end-4) + "_z" + sprintf('%03d', i) + ".png";

                % Convert Image to Double
                data = im2double(image);

                % Write Image
                imwrite(rot90(mat2gray(data(:,:,i))), char(filename));

                % Move Images To Folder
                movefile(char(filename),'png');

                % Increment Counters
                slice_counter = slice_counter + 1;

                percentage = strcat('Please wait. Converting...', ' ', num2str((j/volumes)*100), '% Complete');

                if ((j/volumes)*100) == 100
                    disp('100% Complete!');
                else
                    disp(percentage);
                end                    
            end
            i = i + 1;
        end
   elseif length(nifti_array) ~= 3 || 4
       disp('NIfTI must be 3D or 4D. Please try again.');
   end
end